src/content/content_location.rs

```diff
@@ -0,0 +1,118 @@
+use crate::headers::{HeaderName, HeaderValue, Headers, CONTENT_LOCATION};
+use crate::{Status, Url};
+
+use std::convert::TryInto;
+
+/// Indicates an alternate location for the returned data.
+///
+/// # Specifications
+///
+/// - [RFC 7231, section 3.1.4.2: Content-Location](https://tools.ietf.org/html/rfc7231#section-3.1.4.2)
+///
+/// # Examples
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::{Response,Url};
+/// use http_types::content::{ContentLocation};
+///
+/// let content_location = ContentLocation::new(Url::parse("https://example.net/")?);
+///
+/// let mut res = Response::new(200);
+/// content_location.apply(&mut res);
+///
+/// let content_location = ContentLocation::from_headers(Url::parse("https://example.net/").unwrap(),res)?.unwrap();
+/// assert_eq!(content_location.location(), "https://example.net/");
+/// #
+/// # Ok(()) }
+/// ```
+#[derive(Debug)]
+pub struct ContentLocation {
+    url: Url,
+}
+
+impl ContentLocation {
+    /// Create a new instance of `Content-Location` header.
+    pub fn new(url: Url) -> Self {
+        Self { url }
+    }
+
+    /// Create a new instance from headers.
+    pub fn from_headers<U>(base_url: U, headers: impl AsRef<Headers>) -> crate::Result<Option<Self>>
+    where
+        U: TryInto<Url>,
+        U::Error: std::fmt::Debug,
+    {
+        let headers = match headers.as_ref().get(CONTENT_LOCATION) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        // If we successfully parsed the header then there's always at least one
+        // entry. We want the last entry.
+        let value = headers.iter().last().unwrap();
+        let base = base_url.try_into()?;
```

Comment:
(This probably requires some kind of `map_err()` as noted by the build failures.)

Reply:
Hi @Fishrock123, I use `ok` here and a match later. I don't know if it's the best approach.
Thx!

Reply:
@pepoviola I think what @Fishrock123 meant is that if you call `base_url.try_into()?` then by default a status code of 500 will be set on the error. The right solution would indeed be to call `map_err` and convert the error to a `400` status code. Returning `None` ignores the error, which is less good than having a 500 status code.

Instead the way to address this would be:

```rust
        // If we successfully parsed the header then there's always at least one
        // entry. We want the last entry.
        let value = headers.iter().last().unwrap();
        let base = base_url.try_into().map_err(|mut e| {
            e.set_status(400);
            e
        })?;
```

Hope that's helpful!

Reply:
Hi @yoshuawuyts, thanks for the feedback!! Yes, I wasn't sure how to mapping the errors but now make a lot of sense.
I will change to your approach. Thanks for the help!

Reply:
Hi @yoshuawuyts, sorry to bother you but I changed to the above approach and now I getting this errors related to the Error type and the `set_status` method.

```rust
error[E0599]: no method named `set_status` found for associated type `<U as std::convert::TryInto<url::Url>>::Error` in the current scope
  --> src/content/content_location.rs:58:15
   |
58 |             e.set_status(400);
   |               ^^^^^^^^^^ method not found in `<U as std::convert::TryInto<url::Url>>::Error`

error[E0277]: the trait bound `<U as std::convert::TryInto<url::Url>>::Error: std::error::Error` is not satisfied
  --> src/content/content_location.rs:60:11
   |
60 |         })?;
   |           ^ the trait `std::error::Error` is not implemented for `<U as std::convert::TryInto<url::Url>>::Error`
   |
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `anyhow::Error`
   = note: required because of the requirements on the impl of `std::convert::Into<anyhow::Error>` for `<U as std::convert::TryInto<url::Url>>::Error`
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `error::Error`
   = note: required by `std::convert::From::from`
help: consider further restricting the associated type
   |
46 |         U::Error: std::fmt::Debug, <U as std::convert::TryInto<url::Url>>::Error: std::error::Error
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: `<U as std::convert::TryInto<url::Url>>::Error` cannot be sent between threads safely
  --> src/content/content_location.rs:60:11
   |
60 |         })?;
   |           ^ `<U as std::convert::TryInto<url::Url>>::Error` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `<U as std::convert::TryInto<url::Url>>::Error`
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `anyhow::Error`
   = note: required because of the requirements on the impl of `std::convert::Into<anyhow::Error>` for `<U as std::convert::TryInto<url::Url>>::Error`
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `error::Error`
   = note: required by `std::convert::From::from`
help: consider further restricting the associated type
   |
46 |         U::Error: std::fmt::Debug, <U as std::convert::TryInto<url::Url>>::Error: std::marker::Send
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: `<U as std::convert::TryInto<url::Url>>::Error` cannot be shared between threads safely
  --> src/content/content_location.rs:60:11
   |
60 |         })?;
   |           ^ `<U as std::convert::TryInto<url::Url>>::Error` cannot be shared between threads safely
   |
   = help: the trait `std::marker::Sync` is not implemented for `<U as std::convert::TryInto<url::Url>>::Error`
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `anyhow::Error`
   = note: required because of the requirements on the impl of `std::convert::Into<anyhow::Error>` for `<U as std::convert::TryInto<url::Url>>::Error`
   = note: required because of the requirements on the impl of `std::convert::From<<U as std::convert::TryInto<url::Url>>::Error>` for `error::Error`
   = note: required by `std::convert::From::from`
help: consider further restricting the associated type
   |
46 |         U::Error: std::fmt::Debug, <U as std::convert::TryInto<url::Url>>::Error: std::marker::Sync
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors
```

I try adding the restrictions but don't seems like the best path, right?

Thanks!